File: src/main/kotlin/magicbook/gtlitecore/integration/justenoughitems/JustEnoughItemsModule.java
package magicbook.gtlitecore.integration.justenoughitems;

import magicbook.gtlitecore.api.module.Module;
import magicbook.gtlitecore.api.recipe.frontend.SpacePumpRecipeFrontend;
import magicbook.gtlitecore.api.utils.GTLiteValues;
import magicbook.gtlitecore.api.utils.tuples.Pair;
import magicbook.gtlitecore.common.metatileentity.GTLiteMetaTileEntities;
import magicbook.gtlitecore.core.module.GTLiteModules;
import magicbook.gtlitecore.integration.IntegrationSubModule;
import magicbook.gtlitecore.integration.justenoughitems.category.SpacePumpRecipeCategory;
import magicbook.gtlitecore.integration.justenoughitems.info.SpacePumpRecipeWrapper;
import mezz.jei.api.IGuiHelper;
import mezz.jei.api.IJeiRuntime;
import mezz.jei.api.IModPlugin;
import mezz.jei.api.IModRegistry;
import mezz.jei.api.JEIPlugin;
import mezz.jei.api.ingredients.IIngredientBlacklist;
import mezz.jei.api.ingredients.IIngredientRegistry;
import mezz.jei.api.recipe.IRecipeCategoryRegistration;
import net.minecraftforge.fluids.FluidStack;
import one.util.streamex.EntryStream;
import one.util.streamex.StreamEx;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@JEIPlugin
@Module(moduleId = GTLiteModules.MODULE_JEI,
        containerId = GTLiteValues.MODID,
        modDependencies = { "jei" },
        name = "GregTech Lite JEI Module",
        descriptions = "Just Enough Items (JEI) Module of GregTech Lite Core Mod.")
public class JustEnoughItemsModule extends IntegrationSubModule implements IModPlugin
{

    public static final Logger logger = LogManager.getLogger(GTLiteValues.NAME + " JEI Module");

    private static IJeiRuntime runtime;
    private IIngredientBlacklist blacklist;
    private IGuiHelper guiHelper;
    private IIngredientRegistry ingredientRegistry;

    public JustEnoughItemsModule() {}

    @Override
    public void registerCategories(@NotNull IRecipeCategoryRegistration registry)
    {
        this.guiHelper = registry.getJeiHelpers().getGuiHelper();
        registry.addRecipeCategories(new SpacePumpRecipeCategory(registry.getJeiHelpers().getGuiHelper()));
    }

    @Override
    public void register(@NotNull IModRegistry registry)
    {
        this.blacklist = registry.getJeiHelpers().getIngredientBlacklist();
        this.ingredientRegistry = registry.getIngredientRegistry();

        // SpacePumpRecipeCategory / SpacePumpRecipeWrapper
        String spacePumpId = GTLiteValues.MODID + ":" + "space_pump_module";
        List<SpacePumpRecipeWrapper> spacePumpInfo = StreamEx.of(SpacePumpRecipeFrontend.RECIPES.entrySet())
                .filter(info -> {
                    Pair<Integer, Integer> infoKey = info.getKey();
                    FluidStack infoValue = info.getValue();
                    return infoKey != null && infoKey.getKey() != null && infoKey.getValue() != null
                            && infoValue != null;
                })
                // k1 > k2: (k1, v1), (k2, v1); v1 > v2: (k1, v1), (k1, v2)
                .sorted(Comparator.comparingInt((Map.Entry<Pair<Integer, Integer>, FluidStack> info) -> info.getKey().getKey()) // planetId
                        .thenComparingInt(info -> info.getKey().getValue())) // fluidId
                .map(info -> {
                    Pair<Integer, Integer> infoKey = info.getKey();
                    FluidStack infoValue = info.getValue();
                    return new SpacePumpRecipeWrapper(infoKey.getKey(), infoKey.getValue(), infoValue);
                })
                .distinct()
                .toList();

        registry.addRecipes(spacePumpInfo, spacePumpId);
        registry.addRecipeCatalyst(GTLiteMetaTileEntities.SPACE_PUMP_MK1.getStackForm(), spacePumpId);
        registry.addRecipeCatalyst(GTLiteMetaTileEntities.SPACE_PUMP_MK2.getStackForm(), spacePumpId);
        registry.addRecipeCatalyst(GTLiteMetaTileEntities.SPACE_PUMP_MK3.getStackForm(), spacePumpId);
    }

    @Override
    public void onRuntimeAvailable(@NotNull IJeiRuntime runtime)
    {
        JustEnoughItemsModule.runtime = runtime;
    }

    @NotNull
    @Override
    public Logger getLogger()
    {
        return logger;
    }

}

File: src/main/kotlin/gregtechlite/gtlitecore/integration/justenoughitems/JustEnoughItemsModule.kt
package gregtechlite.gtlitecore.integration.justenoughitems

import gregtech.api.unification.OreDictUnifier
import gregtech.api.unification.material.Materials
import gregtech.api.unification.ore.OrePrefix
import com.morphismmc.morphismlib.util.SidedLogger
import gregtech.api.GTValues
import gregtech.api.GregTechAPI
import gregtech.api.recipes.RecipeMap
import gregtechlite.gtlitecore.api.MOD_ID
import gregtechlite.gtlitecore.api.module.Module
import gregtechlite.gtlitecore.api.recipe.frontend.SpacePumpRecipeFrontend
import gregtechlite.gtlitecore.api.recipe.map.PseudoPairRecipeMap
import gregtechlite.gtlitecore.api.recipe.map.PseudoQuadrupleRecipeMap
import gregtechlite.gtlitecore.api.recipe.map.PseudoTripleRecipeMap
import gregtechlite.gtlitecore.common.metatileentity.GTLiteMetaTileEntities
import gregtechlite.gtlitecore.core.module.GTLiteModules.Companion.MODULE_JEI
import gregtechlite.gtlitecore.integration.IntegrationSubModule
import gregtechlite.gtlitecore.integration.justenoughitems.category.SpacePumpRecipeCategory
import gregtechlite.gtlitecore.integration.justenoughitems.info.SpacePumpRecipeWrapper
import mezz.jei.api.IGuiHelper
import mezz.jei.api.IJeiRuntime
import mezz.jei.api.IModPlugin
import mezz.jei.api.IModRegistry
import mezz.jei.api.JEIPlugin
import mezz.jei.api.ingredients.IIngredientBlacklist
import mezz.jei.api.ingredients.IIngredientRegistry
import mezz.jei.api.recipe.IRecipeCategoryRegistration
import net.minecraftforge.fluids.FluidStack
import org.apache.logging.log4j.Logger

@Suppress("unused")
@JEIPlugin
@Module(moduleId = MODULE_JEI,
        containerId = MOD_ID,
        modDependencies = [ "jei" ],
        name = "GregTech Lite JEI Module",
        descriptions = "Just Enough Items (JEI) Module of GregTech Lite Core Mod.")
class JustEnoughItemsModule : IntegrationSubModule(), IModPlugin
{

    companion object
    {

        @JvmField
        val logger: Logger = SidedLogger("$MOD_ID-jei-module")

    }

    private lateinit var runtime: IJeiRuntime
    private lateinit var blacklist: IIngredientBlacklist
    private lateinit var guiHelper: IGuiHelper
    private lateinit var ingredientRegistry: IIngredientRegistry

    @Suppress("Deprecation", "UnstableApiUsage")
    override fun onRuntimeAvailable(availableRuntime: IJeiRuntime)
    {
        this.runtime = availableRuntime
        logger.info("Registering JEI Blacklist...")

        // Zirconia => Baddeleyite
        blacklist.addIngredientToBlacklist(OreDictUnifier.get(OrePrefix.dust, Materials.Zirconia))
        blacklist.addIngredientToBlacklist(OreDictUnifier.get(OrePrefix.dustSmall, Materials.Zirconia))
        blacklist.addIngredientToBlacklist(OreDictUnifier.get(OrePrefix.dustTiny, Materials.Zirconia))

        // Phosphate is deprecated TODO Remove this blacklist settings in 2.10.
        blacklist.addIngredientToBlacklist(OreDictUnifier.get(OrePrefix.dust, Materials.Phosphate))
        blacklist.addIngredientToBlacklist(OreDictUnifier.get(OrePrefix.dustSmall, Materials.Phosphate))
        blacklist.addIngredientToBlacklist(OreDictUnifier.get(OrePrefix.dustTiny, Materials.Phosphate))
    }
    
    override fun registerCategories(registry: IRecipeCategoryRegistration)
    {
        this.guiHelper = registry.jeiHelpers.guiHelper
        logger.info("Registering JEI Recipe Categories...")

        registry.addRecipeCategories(SpacePumpRecipeCategory(registry.jeiHelpers.guiHelper))
    }
    
    override fun register(registry: IModRegistry)
    {
        this.blacklist = registry.jeiHelpers.ingredientBlacklist
        this.ingredientRegistry = registry.ingredientRegistry
        logger.info("Registering JEI Recipe Wrappers and Catalysts...")

        // SpacePumpRecipeCategory / SpacePumpRecipeWrapper
        val spacePumpId = SpacePumpRecipeCategory.UID
        val spacePumpInfo = SpacePumpRecipeFrontend.RECIPES.entries
            .sortedWith(compareBy<Map.Entry<Pair<Int, Int>, FluidStack>> { it.key.first }.thenBy { it.key.second })
            .map { SpacePumpRecipeWrapper(it.key.first, it.key.second, it.value) }
            .distinct()
            .toList()

        registry.addRecipes(spacePumpInfo, spacePumpId)
        registry.addRecipeCatalyst(GTLiteMetaTileEntities.SPACE_PUMP_MK1.stackForm, spacePumpId)
        registry.addRecipeCatalyst(GTLiteMetaTileEntities.SPACE_PUMP_MK2.stackForm, spacePumpId)
        registry.addRecipeCatalyst(GTLiteMetaTileEntities.SPACE_PUMP_MK3.stackForm, spacePumpId)

        // Add pseudo group mte to its contained recipe map catalyst list.
        for (recipeMap in RecipeMap.getRecipeMaps())
        {
            for (mteRegistry in GregTechAPI.mteManager.registries)
            {
                for (metaTileEntityId in mteRegistry.keys)
                {
                    val metaTileEntity = requireNotNull(mteRegistry.getObject(metaTileEntityId))
                    when (recipeMap)
                    {
                        is PseudoPairRecipeMap -> {
                            registry.addRecipeCatalyst(metaTileEntity.stackForm,
                                                       "${GTValues.MODID}.${recipeMap.leftRecipeMap.unlocalizedName}")
                            registry.addRecipeCatalyst(metaTileEntity.stackForm,
                                                       "${GTValues.MODID}.${recipeMap.rightRecipeMap.unlocalizedName}")
                        }
                        is PseudoTripleRecipeMap -> {
                            registry.addRecipeCatalyst(metaTileEntity.stackForm,
                                                       "${GTValues.MODID}.${recipeMap.leftRecipeMap.unlocalizedName}")
                            registry.addRecipeCatalyst(metaTileEntity.stackForm,
                                                       "${GTValues.MODID}.${recipeMap.middleRecipeMap.unlocalizedName}")
                            registry.addRecipeCatalyst(metaTileEntity.stackForm,
                                                       "${GTValues.MODID}.${recipeMap.rightRecipeMap.unlocalizedName}")
                        }
                        is PseudoQuadrupleRecipeMap -> {
                            registry.addRecipeCatalyst(metaTileEntity.stackForm,
                                                       "${GTValues.MODID}.${recipeMap.firstRecipeMap.unlocalizedName}")
                            registry.addRecipeCatalyst(metaTileEntity.stackForm,
                                                       "${GTValues.MODID}.${recipeMap.secondRecipeMap.unlocalizedName}")
                            registry.addRecipeCatalyst(metaTileEntity.stackForm,
                                                       "${GTValues.MODID}.${recipeMap.thirdRecipeMap.unlocalizedName}")
                            registry.addRecipeCatalyst(metaTileEntity.stackForm,
                                                       "${GTValues.MODID}.${recipeMap.fourthRecipeMap.unlocalizedName}")
                        }
                    }
                }
            }
        }

    }

    override fun getLogger(): Logger = Companion.logger

}
